### PR TITLE
Various fixes

### LINKED ISSUE #1507: Leak in Collection typemaps
The following leaks:
```
import openturns as ot
d = 3
M,N = 10**4,10**5
Norm = ot.Normal(d)  
for i in range(M):
    print(i)
    X = Norm.getSample(N)
    U = X.rank()/(N+1)
    marginals = [ot.KernelSmoothing().build(X[:,k]) for k in range(d)]
    copula = ot.NormalCopulaFactory().build(U)
    distribution = ot.ComposedDistribution(marginals, copula)
```

This could come from the memory allocated in buildCollectionFromPySequence never been freed in 
the "%typemap(in) const CollectionType &" typemap (TypedCollectionInterfaceObject.i), but unsure.

cc @gabrielsarazin

### PR DESCRIPTION
**Pull request by jschueller**
Bugfixes:
- #1377
- #1507
- #1510
- #1521 
- Fix Indices.complement when empty
- Fix ProductCovarianceModel.getMarginal to return marginal wrt output, not input

Enhancements:
- typo in doc, multiform
- FORM: Dont preallocate importance factors

### CODE REVIEW

File: lib/src/Uncertainty/Distribution/KernelSmoothing.cxx [hunk 1]
```diff
@@ -178,9 +178,10 @@ Point KernelSmoothing::computePluginBandwidth(const Sample & sample) const
   const UnsignedInteger dimension = sample.getDimension();
   if (dimension != 1) throw InvalidArgumentException(HERE) << "Error: plugin bandwidth is available only for 1D sample";
   const UnsignedInteger size = sample.getSize();
-  if (size < 2) return Point(1, 0.0);
   // Approximate the derivatives by smoothing under the Normal assumption
   const Scalar sd = sample.computeStandardDeviationPerComponent()[0];
+  if (!(sd > 0.0))
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/Distribution/KernelSmoothing.cxx:183]: in the issue #1515 the sd is > 0

File: lib/src/Uncertainty/Distribution/KernelSmoothing.cxx [hunk 2]
```diff
@@ -235,7 +235,17 @@ Point KernelSmoothing::computeMixedBandwidth(const Sample & sample) const
...
 {
   // For 1D sample, use the rule that give the best tradeoff between speed and precision
-  if (sample.getDimension() == 1) return build(sample, computeMixedBandwidth(sample));
+  if (sample.getDimension() == 1)
+  {
+    try {
+      const Point bandwidth(computeMixedBandwidth(sample));
+      return build(sample, bandwidth);
+    }
+    catch (NotDefinedException &)
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/Distribution/KernelSmoothing.cxx:244]: hmm why a Dirac? what is sd > 0?

Comment (jschueller) [reply lib/src/Uncertainty/Distribution/KernelSmoothing.cxx]: outdated

File: python/doc/examples/reliability_sensitivity/estimate_probability_form.ipynb [hunk 3]
```diff
@@ -146,7 +146,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Create the event whose probability we want to estimate."
+    "Create the event we want to estimate the probability."
```
Comment (josephmure) [inline python/doc/examples/reliability_sensitivity/estimate_probability_form.ipynb:149]:
```suggestion
    "Create the event whose probability we want to estimate."
```
Perhaps you mean "Create the event we want to estimate the probability *of*"? But the original phrase works better in my opinion.

File: python/src/FrankCopulaFactory_doc.i.in [hunk 4]
```diff
@@ -1,7 +1,7 @@
 %feature("docstring") OT::FrankCopulaFactory
 "Frank Copula factory.
 
-The paramters are estimated using the follwing equations:
+The paramters are estimated using the following equations:
```
Comment (josephmure) [inline python/src/FrankCopulaFactory_doc.i.in:4]:
```suggestion
The parameters are estimated using the following equations:
```

File: python/src/GaussianLinearCalibration_doc.i.in [hunk 5]
```diff
@@ -38,12 +38,13 @@ dimension the output dimension of the model, or *global*, ie the matrix describe
...
-The prior distribution of the parameter is a :class:`~openturns.Normal`.  
+The prior distribution of the parameter is a :class:`~openturns.Normal`.
 
-The posterior distribution of the parameter is :class:`~openturns.Normal`. 
+The posterior distribution of the parameter is :class:`~openturns.Normal`.
 
-The resulting error distribution is :class:`~openturns.Normal` and is based on the 
-`errorCovariance` input argument. 
+The resulting error distribution is :class:`~openturns.Normal`
+with a zero mean and with a covariance matrix equal to the
```
Comment (josephmure) [inline python/src/GaussianLinearCalibration_doc.i.in:46]:
```suggestion
with zero mean and with covariance matrix equal to the
```

File: python/src/GaussianNonLinearCalibration_doc.i.in [hunk 6]
```diff
@@ -35,14 +35,15 @@ The posterior distribution of the parameter is :class:`~openturns.Normal`.
...
 This corresponds to using the :class:`~openturns.GaussianLinearCalibration` at the optimum,
-and is named *Laplace approximation* in the bayesian context. 
-However, if the key `GaussianNonLinearCalibration-BootstrapSize` in the :class:`~openturns.ResourceMap` 
-is set to a nonzero positive integer, then a bootstrap resampling of the observations 
-is performed and the posterior distribution is based on a :class:`~openturns.KernelSmoothing` 
-of the sample of boostrap optimum parameters. 
-
-The resulting error distribution is :class:`~openturns.Normal` and is based on the 
-`errorCovariance` input argument. 
+and is named *Laplace approximation* in the bayesian context.
```
Comment (josephmure) [inline python/src/GaussianNonLinearCalibration_doc.i.in:38]:
```suggestion
and is named *Laplace approximation* in the Bayesian context.
```

File: python/src/GaussianNonLinearCalibration_doc.i.in [hunk 7]
```diff
@@ -35,14 +35,15 @@ The posterior distribution of the parameter is :class:`~openturns.Normal`.
...
+However, if the key `GaussianNonLinearCalibration-BootstrapSize` in the :class:`~openturns.ResourceMap`
+is set to a nonzero positive integer, then a bootstrap resampling of the observations
+is performed and the posterior distribution is based on a :class:`~openturns.KernelSmoothing`
+of the sample of boostrap optimum parameters.
+
+The resulting error distribution is :class:`~openturns.Normal`
+with a zero mean and with a covariance matrix equal to the
```
Comment (josephmure) [inline python/src/GaussianNonLinearCalibration_doc.i.in:45]:
```suggestion
with zero mean and with covariance matrix equal to the
```

File: python/src/KernelSmoothing_doc.i.in [hunk 8]
```diff
@@ -327,5 +332,5 @@ bandwidth : :class:`~openturns.Point`
 
 Notes
 ----- 
-Simply use the plugin rule for small sample, and estimate the ratio between the plugin rule and the Silverman rule on a small sample, then scale the Silverman bandwidth computed on the full sample with this ratio.
+Use the plugin rule for small sample, estimate the ratio between the plugin rule and the Silverman rule on this small sample, and finally scale the Silverman bandwidth computed on the full sample with this ratio. The size of the small sample is based on the `KernelSmoothing-SmallSize` key of the `ResourceMap`.
```
Comment (josephmure) [inline python/src/KernelSmoothing_doc.i.in:335]:
```suggestion
Use the plugin rule for a small sample, estimate the ratio between the plugin rule and the Silverman rule on this small sample, and finally scale the Silverman bandwidth computed on the full sample with this ratio. The size of the small sample is based on the `KernelSmoothing-SmallSize` key of the `ResourceMap`.
```

File: python/src/LinearLeastSquaresCalibration_doc.i.in [hunk 9]
```diff
@@ -35,12 +35,12 @@ ensures that the singular values of the covariance matrix are left unmodified.
...
 then all singular values of the covariance matrix are increased by :math:`\epsilon s_1`. 
 
-The resulting distribution of the output error is :class:`~openturns.Normal` with a zero mean 
-and a diagonal covariance matrix computed from the residuals. 
-The residuals are computed based on the linearization of the model, 
-where the Jacobian matrix is evaluated at the candidate point. 
-The diagonal of the diagonal covariance matrix of the output error 
-is contant and is estimated with the unbiased variance estimator. 
+The resulting distribution of the output error is :class:`~openturns.Normal` with a zero mean
+and a diagonal covariance matrix computed from the residuals.
```
Comment (josephmure) [inline python/src/LinearLeastSquaresCalibration_doc.i.in:39]:
```suggestion
The resulting distribution of the output error is :class:`~openturns.Normal` with zero mean.
Its covariance matrix is diagonal and can be computed from the residuals.
```

**Comment by jschueller** [comment]
> @jschueller Please allow me some time to review the rest of your PR.

sorry, I merged without opening my mails first

**Comment by josephmure** [comment]
@jschueller Do you want to discuss the doc corrections I suggested? Or should I submit a pull request that implements them?

**Comment by jschueller** [comment]
a pr would be best

**Pull request merged**